test/cap-automator/index.test.ts

```diff
@@ -0,0 +1,552 @@
+import hre from 'hardhat'
+import { expect } from 'chai'
+import { Web3FunctionHardhat } from '@gelatonetwork/web3-functions-sdk/hardhat-plugin'
+import { SignerWithAddress } from '@nomiclabs/hardhat-ethers/signers'
+import { SnapshotRestorer, impersonateAccount, takeSnapshot, mine } from '@nomicfoundation/hardhat-network-helpers'
+import { Contract } from '@ethersproject/contracts'
+
+import { capAutomatorAbi, erc20Abi, poolAbi, protocolDataProviderAbi } from '../../abis'
+import { addresses } from '../../utils'
+import { Web3FunctionResultCallData } from '@gelatonetwork/web3-functions-sdk/*'
+
+const { w3f, ethers } = hre
+
+describe('CapAutomator', function () {
+    this.timeout(0)
+
+    let snapshotRestorer: SnapshotRestorer
+
+    let capAutomatorW3F: Web3FunctionHardhat
+    let reader: SignerWithAddress
+    let keeper: SignerWithAddress
+
+    let capAutomator: Contract
+    let protocolDataProvider: Contract
+    let pool: Contract
+
+    let sparkAssets: string[]
+
+    const { wbtc, weth, wsteth } = addresses.mainnet
+
+    const wbtcWhale = '0xb20Fb60E27a1Be799b5e04159eC2024CC3734eD7' as const
+    const wstethWhale = '0x5fEC2f34D80ED82370F733043B6A536d7e9D7f8d' as const
+
+    const deposit = async (signerAddress: string, tokenAddress: string, amount: BigInt) => {
+        await impersonateAccount(signerAddress)
+        const signer = await hre.ethers.getSigner(signerAddress)
+
+        const token = new Contract(tokenAddress, erc20Abi, signer)
+
+        await token.approve(addresses.mainnet.pool, amount)
+
+        const pool = new Contract(addresses.mainnet.pool, poolAbi, signer)
+        await pool.supply(tokenAddress, amount, signerAddress, 0)
+    }
+
+    const borrow = async (signerAddress: string, tokenAddress: string, amount: BigInt) => {
+        await impersonateAccount(signerAddress)
+        const signer = await hre.ethers.getSigner(signerAddress)
+
+        const pool = new Contract(addresses.mainnet.pool, poolAbi, signer)
+        await pool.borrow(tokenAddress, amount, 2, 0, signerAddress)
+    }
+
```

Comment:
Lets add withdraw and repay functions here and get coverage for the cap reduction cases, can basically just take existing tests and do them the other way